cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,880 @@
+"""
+Copyright 2023 Impulse Innovations Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from __future__ import annotations
+
+import logging
+from copy import deepcopy
+from functools import wraps
+from typing import Any, Dict, List, Optional, Tuple, Union, cast
+
+import numpy
+
+from cai_causal_graph import CausalGraph
+from cai_causal_graph.exceptions import CausalGraphErrors
+from cai_causal_graph.graph_components import Edge, Node
+from cai_causal_graph.interfaces import HasIdentifier, HasMetadata
+from cai_causal_graph.type_definitions import EDGE_T, TIME_LAG, VARIABLE_NAME, NodeLike, NodeVariableType
+from cai_causal_graph.utils import get_name_with_lag, get_variable_name_and_lag
+
+logger = logging.getLogger(__name__)
+
+# TODO: we can do one general for many other things as well
```

Comment:
Is there a JIRA ticket associated with this? Should we add JIRA ticket numbers on public repos, @andrew-causalens? I honestly would rather add a JIRA ticket directly and then remove this TODO altogether, since it's public and all.